tidy3d/components/device/source.py

```diff
@@ -0,0 +1,81 @@
+"""Defines heat material specifications"""
+from __future__ import annotations
+
+from abc import ABC
+from typing import Union, Tuple
+
+import pydantic.v1 as pd
+
+from .viz import plot_params_heat_source
+
+from ..base import cached_property
+from ..base_sim.source import AbstractSource
+from ..data.data_array import TimeDataArray
+from ..viz import PlotParams
+
+from ...constants import VOLUMETRIC_HEAT_RATE, PERCMCUBE
+
+
+class DeviceSource(AbstractSource, ABC):
+    """Abstract device source."""
+
+    structures: Tuple[str, ...] = pd.Field(
+        title="Target Structures",
+        description="Names of structures where to apply heat source.",
+    )
+
+    @cached_property
+    def plot_params(self) -> PlotParams:
+        """Default parameters for plotting a Source object."""
+        return plot_params_heat_source
+
+
+class UniformHeatSource(DeviceSource):
+    """Volumetric heat source.
+
+    Example
+    -------
+    >>> heat_source = UniformHeatSource(rate=1, structures=["box"])
+    """
+
+    rate: Union[float, TimeDataArray] = pd.Field(
+        title="Volumetric Heat Rate",
+        description="Volumetric rate of heating or cooling (if negative) in units of "
+        f"{VOLUMETRIC_HEAT_RATE}.",
+        units=VOLUMETRIC_HEAT_RATE,
+    )
+
+
+class ExpressionHeatSource(DeviceSource):
+    """Expression heat source
+
+    Example
+    -------
+    >>> r = 1e-3
+    >>> heat_source = ExpressionHeatSource(expression=f"{r}*(sqrt(x) + sin(y)*pow(z,3))", structures=["box"])
+    """
+
+    expression: str = pd.Field(
```

Comment:
Can remain `rate:` probably. Is there a way to validate that provided expression is valid? it could be not a great experience having to submit a simulation without knowing whether the source will error or not

Reply:
hmmm that's a good point. For what I can see in the C++ code the expression needs to evaluate to a valid C++ code, for which I believe they use a 3rd party library. So not sure if we should keep it and warn users we won't check the expression before submitting; or just get rid of this option.